lib/BloqadeQMC/src/ising/Rydberg.jl

```diff
@@ -75,80 +88,91 @@ end
+#             if length(uniq_dist) == 0
+#                 push!(uniq_dist, _dist[i])
+#             elseif !(last(uniq_dist) ≈ _dist[i])
+#                 push!(uniq_dist, _dist[i])
+#             end
+#         end
+#         smallest_k = sort!(uniq_dist)[1:(trunc+1)]
+#         dist = copy(lat.distance_matrix)
+#         for i in eachindex(dist)
+#             if dist[i] > last(smallest_k) && !(dist[i] ≈ last(smallest_k))
+#                 dist[i] = zero(dist[i])
+#             end
+#         end
+#     elseif lat isa Rectangle && all(lat.PBC)
+#         V = zeros(Ns, Ns)
+#         K = 3  # figure out an efficient way to set this dynamically
+
+#         dist = zeros(Ns, Ns)
+#         for v2 in -K:K, v1 in -K:K
+#             dV = zeros(Ns, Ns)
+#             for x2 in axes(dV, 2), x1 in axes(dV, 1)
+#                 i1, j1 = divrem(x1 - 1, lat.n1)
+#                 i2, j2 = divrem(x2 - 1, lat.n1)
+#                 r = [i2 - i1 + v1*lat.n1, j2 - j1 + v2*lat.n2]
+#                 dV[x1, x2] += Ω * (R_b/norm(r, 2))^6
+#             end
+#             # @show v2, v1, maximum(abs, dV)
+
+#             V += dV
+#         end
+
+#         V = (V + V') / 2  # should already be symmetric but just in case
+
+#         return Rydberg(UpperTriangular(triu!(V, 1)), Ω*ones(Ns), δ*ones(Ns), lat; epsilon=epsilon)
+#     else
+#         dist = lat.distance_matrix
+#     end
+
+#     @inbounds for i in 1:(Ns-1)
+#         for j in (i+1):Ns
+#             # a zero entry in distance_matrix means there should be no bond
+#             V[i, j] = dist[i, j] != 0.0 ? Ω * (R_b / dist[i, j])^6 : 0.0
+#         end
+#     end
+#     V = UpperTriangular(triu!(V, 1))
+
+#     return Rydberg(V, Ω*ones(Ns), δ*ones(Ns), lat; epsilon=epsilon)
+# end
 
-    @inbounds for i in 1:(Ns-1)
-        for j in (i+1):Ns
-            # a zero entry in distance_matrix means there should be no bond
-            V[i, j] = dist[i, j] != 0.0 ? Ω * (R_b / dist[i, j])^6 : 0.0
-        end
-    end
-    V = UpperTriangular(triu!(V, 1))
+# function Rydberg(V::AbstractMatrix{T}, Ω::AbstractVector{T}, δ::AbstractVector{T}, lattice::Lattice; epsilon=zero(T)) where T
+#     ops, p, energy_shift = make_prob_vector(AbstractRydberg, V, Ω, δ, epsilon=epsilon)
+#     op_sampler = ImprovedOperatorSampler(AbstractLTFIM, ops, p)
+#     return Rydberg{typeof(op_sampler), typeof(V), typeof(Ω), typeof(δ), typeof(lattice)}(op_sampler, V, Ω, δ, lattice, energy_shift)
+# end
 
-    return Rydberg(V, Ω*ones(Ns), δ*ones(Ns), lat; epsilon=epsilon)
-end
+total_hx(H::Rydberg)::Float64 = sum(H.Ω) / 2
+haslongitudinalfield(H::AbstractRydberg) = !iszero(H.δ)
 
-function Rydberg(V::AbstractMatrix{T}, Ω::AbstractVector{T}, δ::AbstractVector{T}, lattice::Lattice; epsilon=zero(T)) where T
-    ops, p, energy_shift = make_prob_vector(AbstractRydberg, V, Ω, δ, epsilon=epsilon)
+function rydberg_QMC(atoms::AtomList{1, Float64}; C = 2π * 862690, Ω::Float64, Δ::Float64)  # intended to mimic the rydberg_h interface 
```

Comment:
The code supports non-uniform Delta and Omega, we should have those be supported as well. Lets chat about that in person tomorrow.

Reply:
Yup, was thinking about that, too. Thought I'd first implement the vanilla version and then chat with you :+1: